wireguard/metrics.go

```diff
@@ -0,0 +1,227 @@
+// Copyright (c) 2021 Tigera, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package wireguard
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/projectcalico/felix/netlinkshim"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/sirupsen/logrus"
+	"golang.zx2c4.com/wireguard/wgctrl/wgtypes"
+)
+
+const (
+	labelHostname      = "hostname"
+	labelPublicKey     = "public_key"
+	labelInterfaceName = "iface"
+	labelListenPort    = "listen_port"
+	labelPeerKey       = "peer_key"
+	labelPeerEndpoint  = "peer_endpoint"
+)
+
+var _ prometheus.Collector = (*Metrics)(nil)
+
+const (
+	wireguardMetaFQName   = "wireguard_meta"
+	wireguardMetaHelpText = "wireguard interface and runtime metadata"
+
+	wireguardLatestHandshakeIntervalFQName   = "wireguard_latest_handshake_seconds"
+	wireguardLatestHandshakeIntervalHelpText = "wireguard interface latest handshake unix timestamp in seconds to a peer"
+
+	wireguardBytesSentFQName   = "wireguard_bytes_sent_total"
+	wireguardBytesSentHelpText = "wireguard interface total outgoing bytes to peer"
+
+	wireguardBytesRcvdFQName   = "wireguard_bytes_rcvd_total"
+	wireguardBytesRcvdHelpText = "wireguard interface total incoming bytes to peer"
+)
+
+func init() {
+	prometheus.MustRegister(
+		MustNewWireguardMetrics(),
+	)
+}
+
+type Metrics struct {
```

Comment:
What's the reason for implementing your own metric (rather than using off-the-shelf ones from the prometheus client lib)?

Reply:
I wanted _something_ to trigger stats collection. So a convenient way is just to hook it in Collect(). Otherwise, I'd have to find another way e.g. a timer or an event. Now, I could've used CounterFunc but considering the calls to the data plane that each metric requires it could get quite complicated to synchronize. I hoped that this would be way simpler!